src/poll.rs

```diff
@@ -1,60 +1,231 @@
+                    Err(e) => break Poll::Ready(Err(e.into())),
+                }
             }
+        } else {
+            Poll::Ready(Ok(read_buffer.pop_front().unwrap()))
         }
     }
 
-    fn recv_message(&self, msg: &mut zmq::Message) -> Poll<(), Error> {
-        let ready = Ready::readable();
+    /// Attempt to receive a Multipart message from a ZeroMQ socket.
+    ///
+    /// Either the whole multipart at once or nothing is received (according to
+    /// [http://zguide.zeromq.org/page:all#Multipart-Messages], when one part of a multipart message
+    /// has been received, all the others are already available as well).
+    ///
+    /// If nothing was received, the read flag is cleared.
+    pub(crate) fn multipart_recv(&self, cx: &mut Context<'_>) -> Poll<Result<Multipart>> {
+        ready!(self.multipart_poll_read_ready(cx))?;
 
-        match self.poll_read_ready(ready)? {
-            Async::Ready(_) => {
-                if let Err(e) = self.get_ref().io.recv(msg, zmq::DONTWAIT) {
-                    if e == zmq::Error::EAGAIN {
-                        self.clear_read_ready(ready)?;
-                        return Ok(Async::NotReady);
-                    } else {
-                        return Err(e.into());
+        let mut buffer = Multipart::default();
+        loop {
+            let mut msg = zmq::Message::new();
+            match self.get_socket().recv(&mut msg, zmq::DONTWAIT) {
+                Ok(_) => {
+                    let more = msg.get_more();
+                    buffer.push_back(msg);
+                    if !more {
+                        break;
                     }
                 }
+                Err(zmq::Error::EAGAIN) => {
+                    assert!(buffer.is_empty());
+                    log::warn!("EAGAIN during first message read");
+                    self.clear_read_ready(cx, Ready::readable())?;
+                    return Poll::Pending;
+                }
+                Err(e) => return Poll::Ready(Err(e.into())),
+            }
+        }
+
+        assert!(!buffer.is_empty());
+        Poll::Ready(Ok(buffer))
+    }
 
-                return Ok(Async::Ready(()));
+    /// Attempt to send a multipart message.
+    ///
+    /// Sending the whole message at once may not be possible.
+    /// If the function returns `Poll::Ready(Ok(()))`, the whole message has been sent.
+    /// If the function returns `Poll::Pending`, the remaining message in the buffer should be
+    /// attempted to be written the next time the socket is polled.
+    pub(crate) fn multipart_send(&self, buffer: &mut Multipart) -> Poll<Result<()>> {
+        let len = buffer.len();
+
+        while let Some(msg) = buffer.pop_front() {
+            let mut flags = zmq::DONTWAIT;
+            if !buffer.is_empty() {
+                flags |= zmq::SNDMORE;
             }
-            Async::NotReady => {
-                return Ok(Async::NotReady);
+
+            match self.get_socket().send(&*msg, flags) {
```

Comment:
We are already sending the messages by reference, so for EAGAIN handling can't we just retain all parts of the message until the whole thing is sent?  There's nothing really gained by popping from the vecdeque rather than just referencing by index, is there?

Reply:
Right, I remember now what's the problem, I have dealt with it Right, I remember now what's the problem, I have dealt with it in the beginning of this PR.
in the beginning of this PR.
So what happens here is actually that the message gets copied here, I forgot about this.

`*msg` derefs into `[u8]`.
`&*msg` turns it into `&[u8]`
`send` takes `&[u8]` and converts it into a brand new `zmq::Message`.

Since we already copy the message, you're right that we could actually keep the whole `VecDeque` and start from the beginning if we find `EAGAIN`. However, I'm still not sure if that's the correct thing to do (the Github issue reference CZMQ implementation, but I don't see that it would actually retry the whole message in case of EAGAIN, so maybe we can wait for further clarification).

If tokio's `poll_write` would reliably work (not sure if that's true now for our usage), we could ignore EAGAIN during the first message send. If we then also ignored EAGAIN in the middle of the multipart (not sure if we can do that), we could actually avoid a copy altogether. But for now it's probably safer to keep the copy there.

Reply:
I would hope that between `tmq` and `zmq` and `libzmq` we could do with one copy of the message, instead of the 3 copies we have today.   I think that's beyond the scope here, so my only concern is that we not craft the API in such a way that it requires a copy in the tmq layer. 

I think the best route might be to do what you suggested earlier: in the event of a EAGAIN in the middle of a multipart, don't try to resend.  Don't keep sent parts.  Push an `Err` back to the client, and let them decide what to do.  If they need to resend, then push that allocation onto the client.

Eventually, I can imagine that the Message would get pushed in a zero-copy fashion all the way down into libzmq (which, via custom message destructor callbacks, can assume ownership of client memory).

Reply:
Zero copy is of course the goal here, although I'm not sure if it's truly possible, because the ZMQ polling interface is pretty weird and undocumented. We can try to put a `panic!` in the case of EAGAIN during send and see if it happens in real world usage. We consume the input messages anyway, so changing zero copy to copy later would not be a breaking change.

Reply:
Great, sounds like we're in agreement.  I slightly prefer an error over `panic!` as it would give the crate's clients control of how to react.  In any case, thanks for talking through this! 👍

Reply:
Error is better of course, but there should be an assert that EAGAIN didn't occur after the first send. That would mean that there is an error in our  `poll` logic.